mqtt-streaming/src/main/scala/akka/stream/alpakka/mqtt/streaming/impl/ServerState.scala

```diff
@@ -0,0 +1,800 @@
+                    .onFailure[Unpublisher.UnsubscribeFailed.type](SupervisorStrategy.stop.withLoggingEnabled(false)),
+                  unpublisherName
+                )
+                context.watchWith(unpublisher, UnpublisherFree(unsubscribe.topicFilters))
+              case _: Some[_] => // Ignored for existing unsubscriptions
+            }
+            Behaviors.same
+          case (_, UnpublisherFree(topicFilters)) =>
+            val unsubscribedTopicFilters =
+              data.publishers.filter(publisher => topicFilters.exists(matchTopicFilter(_, publisher)))
+            clientConnected(data.copy(publishers = data.publishers -- unsubscribedTopicFilters))
+          case (_, PublishReceivedFromRemote(publish, local))
+              if (publish.flags & ControlPacketFlags.QoSReserved).underlying == 0 =>
+            local ! Consumer.ForwardPublish
+            Behaviors.same
+          case (context, PublishReceivedFromRemote(publish @ Publish(_, topicName, Some(packetId), _), local)) =>
+            val consumerName = ActorName.mkName(ConsumerNamePrefix + topicName + packetId.underlying)
+            context.child(consumerName) match {
+              case None =>
+                context.spawn(
+                  Behaviors
+                    .supervise(Consumer(publish, packetId, local, data.consumerPacketRouter, data.settings))
+                    .onFailure[Consumer.ConsumeFailed.type](SupervisorStrategy.stop.withLoggingEnabled(false)),
+                  consumerName
+                )
+              case _: Some[_] => // Ignored for existing consumptions
+            }
+            Behaviors.same
+          case (_, PublishReceivedLocally(publish, _))
+              if (publish.flags & ControlPacketFlags.QoSReserved).underlying == 0 &&
+              data.publishers.exists(matchTopicFilter(_, publish.topicName)) =>
+            data.remote.offer(ForwardPublish(publish, None))
+            clientConnected(data)
+          case (context, prl @ PublishReceivedLocally(publish, publishData))
+              if data.publishers.exists(matchTopicFilter(_, publish.topicName)) =>
+            val producerName = ActorName.mkName(ProducerNamePrefix + publish.topicName)
+            context.child(producerName) match {
+              case None if !data.pendingLocalPublications.exists(_._1 == publish.topicName) =>
+                context.watchWith(
+                  context.spawn(Producer(publish, publishData, selfAsRemote, data.producerPacketRouter, data.settings),
+                                producerName),
+                  ProducerFree(publish.topicName)
+                )
+                clientConnected(data)
+              case _ =>
+                clientConnected(
+                  data.copy(pendingLocalPublications = data.pendingLocalPublications :+ (publish.topicName -> prl))
+                )
+            }
```

Comment:
code style: some of these bigger actors might benefit from extracting the code into smaller methods, instead of putting everything in one big pattern match. I think that improves readability and navigation.

Reply:
I'm on the fence on that one... Let's consult the master of style - @hseeberger ? :-)